File: web/src/mock/fixtures/lineage.ts
import type { LineageGraph } from "@/types/lineage";

// A graph around `analytics.daily_active_users`, exercising every node kind the
// UI has to render: plain tables, an imported external source, and a node the
// viewer holds no grant on. Two providers agree on one edge, which is what the
// merged `providers` list is for — and on that edge the imported claim has gone
// stale while the execution-derived one has not, so the per-provider freshness
// markers are reachable in dev without a handler override.
function key(schema: string, table: string): string {
  return `cat:11111111-1111-1111-1111-111111111111/${schema}/${table}`;
}

const NOW = "2026-08-14T09:30:00Z";
const EARLIER = "2026-07-01T08:00:00Z";
// Comfortably past the default 30-day staleness window.
const LONG_AGO = "2026-02-11T08:00:00Z";

export function makeLineage(): LineageGraph {
  return {
    root: key("analytics", "daily_active_users"),
    nodes: [
      {
        key: "ext:crm_pg/public/customers",
        kind: "external",
        catalog: null,
        schema_name: "public",
        table: "customers",
        system: "crm_pg",
        distance: -2,
      },
      {
        key: key("raw", "events"),
        kind: "table",
        catalog: "acme_analytics",
        schema_name: "raw",
        table: "events",
        system: null,
        distance: -1,
      },
      {
        key: "redacted:9f2c4a1b7e0d3856",
        kind: "redacted",
        catalog: null,
        schema_name: null,
        table: null,
        system: null,
        distance: -1,
      },
      {
        key: key("analytics", "daily_active_users"),
        kind: "table",
        catalog: "acme_analytics",
        schema_name: "analytics",
        table: "daily_active_users",
        system: null,
        distance: 0,
      },
      {
        key: key("analytics", "funnel"),
        kind: "table",
        catalog: "acme_analytics",
        schema_name: "analytics",
        table: "funnel",
        system: null,
        distance: 1,
      },
    ],
    edges: [
      {
        // Nothing has re-imported this in months: stale, and the only producer
        // asserting it, so the edge itself is stale.
        source_key: "ext:crm_pg/public/customers",
        target_key: key("raw", "events"),
        operation: "model",
        providers: [
          {
            name: "dbt",
            first_seen_at: EARLIER,
            last_seen_at: LONG_AGO,
            observation_count: 12,
            stale: true,
            column_lineage: "unsupported",
          },
        ],
        confidence: "exact",
        first_seen_at: EARLIER,
        last_seen_at: LONG_AGO,
        observation_count: 12,
        stale: true,
        last_query_id: null,
        // An external source whose columns cannot be tied to an asset: data does
        // flow, we just cannot say which columns, and that is not the same as
        // saying none do.
        columns: [],
        column_lineage: "unsupported",
      },
      {
        // Two producers, one of which stopped. The edge stays current because
        // the other still confirms it.
        source_key: key("raw", "events"),
        target_key: key("analytics", "daily_active_users"),
        operation: "create_table_as",
        providers: [
          {
            name: "dbt",
            first_seen_at: EARLIER,
            last_seen_at: LONG_AGO,
            observation_count: 12,
            stale: true,
            column_lineage: "derived",
          },
          {
            name: "execution",
            first_seen_at: EARLIER,
            last_seen_at: NOW,
            observation_count: 35,
            stale: false,
            column_lineage: "derived",
          },
        ],
        confidence: "exact",
        first_seen_at: EARLIER,
        last_seen_at: NOW,
        observation_count: 47,
        stale: false,
        last_query_id: "q-1",
        // Two upstream columns feeding one output, and one feeding two — both
        // shapes the model has to support — plus a mapping that has gone stale
        // on its own while the rest of the edge stayed current.
        columns: [
          {
            source_column: "user_id",
            target_column: "user_id",
            providers: ["dbt", "execution"],
            stale: false,
          },
          {
            source_column: "event_id",
            target_column: "active_count",
            providers: ["execution"],
            stale: false,
          },
          {
            source_column: "session_id",
            target_column: "active_count",
            providers: ["execution"],
            stale: false,
          },
          {
            source_column: "occurred_at",
            target_column: "day",
            providers: ["dbt", "execution"],
            stale: true,
          },
        ],
        column_lineage: "derived",
      },
      {
        source_key: "redacted:9f2c4a1b7e0d3856",
        target_key: key("analytics", "daily_active_users"),
        operation: "create_table_as",
        providers: [
          {
            name: "execution",
            first_seen_at: EARLIER,
            last_seen_at: NOW,
            observation_count: 47,
            stale: false,
            column_lineage: "derived",
          },
        ],
        confidence: "exact",
        first_seen_at: EARLIER,
        last_seen_at: NOW,
        observation_count: 47,
        stale: false,
        last_query_id: "q-1",
        // Withheld: naming a restricted table's columns would give away more
        // than the table name the redaction already holds back.
        columns: [],
        column_lineage: "derived",
      },
      {
        source_key: key("analytics", "daily_active_users"),
        target_key: key("analytics", "funnel"),
        operation: "insert",
        providers: [
          {
            name: "execution",
            first_seen_at: EARLIER,
            last_seen_at: NOW,
            observation_count: 3,
            stale: false,
            column_lineage: "derived",
          },
        ],
        confidence: "exact",
        first_seen_at: EARLIER,
        last_seen_at: NOW,
        observation_count: 3,
        stale: false,
        last_query_id: "q-2",
        // Read and filtered on, but none of its values reach the target. The
        // finding the table graph could never state.
        columns: [],
        column_lineage: "derived",
      },
    ],
    truncated: false,
    hidden: false,
    columns_truncated: false,
  };
}

// A table nothing has been built from and nothing reads: the empty state.
export function makeEmptyLineage(root: string): LineageGraph {
  return {
    root,
    nodes: [],
    edges: [],
    truncated: false,
    hidden: false,
    columns_truncated: false,
  };
}

// A table that *does* have lineage, all of it in catalogs this workspace cannot
// see. Indistinguishable from the empty graph without the `hidden` flag, which
// is the entire reason the flag exists.
export function makeHiddenLineage(root: string): LineageGraph {
  return {
    root,
    nodes: [],
    edges: [],
    truncated: false,
    hidden: true,
    columns_truncated: false,
  };
}

export let LINEAGE = makeLineage();

export function resetLineage(): void {
  LINEAGE = makeLineage();
}

File: api/src/api/models/lineage.py
"""The lineage graph: which datasets were produced from which other datasets.

A :class:`LineageEdge` is a single ``source -> target`` relationship, meaning
"data in the target was produced using the source". Edges are stored denormalized
— each row carries both endpoints' names directly — for the same reason
``catalog_grants`` and ``table_metadata`` do: Polaris owns catalog structure and
the control plane never caches it (I3). There is deliberately no asset table.

Edges are keyed by *catalog*, not workspace, mirroring
:class:`~api.models.table_metadata.TableMetadata`'s rationale: a relationship
between two tables is an intrinsic fact about the data, so a catalog attached to
both ``dev`` and ``prod`` has one lineage graph rather than two. ``workspace_id``
records where the relationship was *observed* and is provenance only — the read
boundary is enforced at query time by clamping traversal to the catalogs the
requesting workspace attaches.

Every edge names the ``provider`` that asserted it, and the provider is part of
the identity key. Two producers describing the same pair therefore coexist as two
rows rather than overwriting each other, and reconciliation is always scoped to a
single provider. Disagreement between producers is information, not a conflict to
resolve; the read API merges the rows into one graph edge listing both.
"""

from __future__ import annotations

import uuid
from datetime import datetime

from sqlalchemy import (
    DateTime,
    ForeignKey,
    Index,
    Integer,
    String,
    UniqueConstraint,
    func,
)
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column

from api.db.base import Base


class LineageEdge(Base):
    """One ``source -> target`` dataset relationship, as asserted by one provider.

    The ``*_key`` columns are the canonical asset key (see
    ``api.services.lineage.keys``): a single indexed string per endpoint, built on
    ``catalogs.id`` rather than the catalog slug so renaming a catalog does not
    orphan its lineage. The structured ``*_catalog_id``/``*_schema``/``*_table``
    columns are kept alongside for display and for cascade cleanup, not as the
    join key.

    An endpoint with ``catalog_id IS NULL`` is *external* — something outside any
    DuckHaven catalog, named by the producer (``system``). That is how an imported
    graph keeps its roots instead of silently losing them.
    """

    __tablename__ = "lineage_edges"
    __table_args__ = (
        # The dedup key. `provider` is part of it on purpose: it is what lets
        # execution-derived and imported lineage assert the same pair without
        # either one clobbering the other.
        UniqueConstraint("provider", "source_key", "target_key", name="uq_lineage_edges_identity"),
        Index("ix_lineage_edges_source_key", "source_key"),
        Index("ix_lineage_edges_target_key", "target_key"),
        Index("ix_lineage_edges_provider_run", "provider", "provider_run_id"),
        # Cleanup when a table or schema is dropped.
        Index(
            "ix_lineage_edges_source_catalog",
            "source_catalog_id",
            "source_schema",
            "source_table",
        ),
        Index(
            "ix_lineage_edges_target_catalog",
            "target_catalog_id",
            "target_schema",
            "target_table",
        ),
    )

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)

    source_key: Mapped[str] = mapped_column(String(1024), nullable=False)
    source_catalog_id: Mapped[uuid.UUID | None] = mapped_column(
        ForeignKey("catalogs.id", ondelete="CASCADE"), nullable=True
    )
    # Set exactly when `source_catalog_id` is NULL: the external system's name.
    source_system: Mapped[str | None] = mapped_column(String(128), nullable=True)
    source_schema: Mapped[str] = mapped_column(String(255), nullable=False)
    source_table: Mapped[str] = mapped_column(String(255), nullable=False)

    target_key: Mapped[str] = mapped_column(String(1024), nullable=False)
    target_catalog_id: Mapped[uuid.UUID | None] = mapped_column(
        ForeignKey("catalogs.id", ondelete="CASCADE"), nullable=True
    )
    target_system: Mapped[str | None] = mapped_column(String(128), nullable=True)
    target_schema: Mapped[str] = mapped_column(String(255), nullable=False)
    target_table: Mapped[str] = mapped_column(String(255), nullable=False)

    # Who asserted this relationship: "execution" for lineage derived from SQL
    # DuckHaven ran, or an importer's name (e.g. "dbt").
    provider: Mapped[str] = mapped_column(String(50), nullable=False)
    # The import batch this edge last arrived in — the unit of reconciliation.
    # Null for execution-derived edges, which are never batch-reconciled.
    provider_run_id: Mapped[str | None] = mapped_column(String(255), nullable=True)
    # Where the relationship was observed. Provenance only: visibility is decided
    # by which catalogs the *requesting* workspace attaches, not by this column.
    workspace_id: Mapped[uuid.UUID | None] = mapped_column(
        ForeignKey("workspaces.id", ondelete="SET NULL"), nullable=True
    )
    # What kind of statement produced it: create_table_as, create_view, insert,
    # update, merge, delete — or an importer's own vocabulary (dbt: "model").
    operation: Mapped[str | None] = mapped_column(String(32), nullable=True)
    # "exact" when parsed or declared; "inferred" is reserved for future heuristic
    # producers. Nothing emits "inferred" today — the column exists so uncertainty
    # has somewhere to live rather than being laundered into fact.
    confidence: Mapped[str] = mapped_column(String(16), nullable=False, default="exact")
    # Whether this provider worked out the column-level detail, and not merely
    # whether any :class:`LineageColumnEdge` rows exist. "derived" with no children
    # is a real and useful answer — the source was joined against or filtered on but
    # none of its values reached the target — and it must not read as "we don't
    # know", which is what "unknown" (never attempted) and "unsupported" (attempted,
    # could not) say.
    column_lineage: Mapped[str] = mapped_column(
        String(16), nullable=False, default="unknown", server_default="unknown"
    )

    # The query that most recently produced this edge, for click-through to the
    # exact SQL. A soft pointer with no FK (like `schedules.last_run_query_id`) so
    # trimming query history never cascades into the graph.
    last_query_id: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), nullable=True)

    first_seen_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, server_default=func.now()
    )
    last_seen_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, server_default=func.now()
    )
    # How many times this relationship has been re-asserted. Together with the
    # two timestamps this answers "when did this start / last happen / how often"
    # without an event log — the history itself stays recoverable from `queries`,
    # which retains every statement's SQL text indefinitely.
    observation_count: Mapped[int] = mapped_column(Integer, nullable=False, default=1)


class LineageColumnEdge(Base):
    """A column-level refinement of a :class:`LineageEdge`.

    Deliberately a child of the dataset edge rather than an independent
    relationship: a column edge cannot exist without the dataset edge it refines,
    so the table graph is always a correct coarsening of the column graph. Hanging
    off ``edge_id`` also means these inherit the parent's provenance for free —
    which provider asserted the mapping is whichever one owns the parent row — and
    that a table rename, which rewrites the parent in place, carries them along
    without touching a single child.

    Re-observation *accumulates*. Two statements can legitimately build the same
    target from the same source through different columns, so the union is the
    truth and deleting whatever the newest statement did not mention would make a
    mapping flap in and out. A mapping nothing re-asserts ages out through
    ``last_seen_at`` instead, exactly as the parent edge does.
    """

    __tablename__ = "lineage_column_edges"
    __table_args__ = (
        UniqueConstraint(
            "edge_id", "source_column", "target_column", name="uq_lineage_column_edges_identity"
        ),
        Index("ix_lineage_column_edges_edge", "edge_id"),
        # "What feeds this column?" The unique constraint above already indexes
        # the (edge_id, source_column) prefix for the other direction.
        Index("ix_lineage_column_edges_edge_target", "edge_id", "target_column"),
    )

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    edge_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("lineage_edges.id", ondelete="CASCADE"), nullable=False
    )
    source_column: Mapped[str] = mapped_column(String(255), nullable=False)
    target_column: Mapped[str] = mapped_column(String(255), nullable=False)

    first_seen_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, server_default=func.now()
    )
    last_seen_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, server_default=func.now()
    )
